# Machine Learning in R: part 2

## This week focuses on improvement. We will our data cleaning code more efficient and then use some tuning methods to improve the predictive ability of models we build.


## 1a. Cleaning and formatting the data (from last week)

Below is the code from the loading, cleaning and train/test split sections we went over last week. This is all we requre to get the data into the format needed to being training some machine learning models.

In [75]:
library(tidyverse)

housing = read.csv('./housing.csv')

housing$total_bedrooms[is.na(housing$total_bedrooms)] = median(housing$total_bedrooms , na.rm = TRUE)

housing$mean_bedrooms = housing$total_bedrooms/housing$households
housing$mean_rooms = housing$total_rooms/housing$households

drops = c('total_bedrooms', 'total_rooms')

housing = housing[ , !(names(housing) %in% drops)]

categories = unique(housing$ocean_proximity)
#split the categories off
cat_housing = data.frame(ocean_proximity = housing$ocean_proximity)

for(cat in categories){
    cat_housing[,cat] = rep(0, times= nrow(cat_housing))
}

for(i in 1:length(cat_housing$ocean_proximity)){
    cat = as.character(cat_housing$ocean_proximity[i])
    cat_housing[,cat][i] = 1
}

cat_columns = names(cat_housing)
keep_columns = cat_columns[cat_columns != 'ocean_proximity']
cat_housing = select(cat_housing,one_of(keep_columns))
drops = c('ocean_proximity','median_house_value')
housing_num =  housing[ , !(names(housing) %in% drops)]


scaled_housing_num = scale(housing_num)

cleaned_housing = cbind(cat_housing, scaled_housing_num, median_house_value=housing$median_house_value)


set.seed(19) # Set a random seed so that same sample can be reproduced in future runs

sample = sample.int(n = nrow(cleaned_housing), size = floor(.8*nrow(cleaned_housing)), replace = F)
train = cleaned_housing[sample, ] #just the samples
test  = cleaned_housing[-sample, ] #everything but the samples


train_y = train[,'median_house_value']
train_x = train[, names(train) !='median_house_value']

test_y = test[,'median_house_value']
test_x = test[, names(test) !='median_house_value']

head(train)


NEAR BAY <1H OCEAN INLAND NEAR OCEAN ISLAND longitude   latitude  
2418  0        0         1      0          0       0.06473791  0.4485767
9990  0        0         1      0          0      -0.74882545  1.6471053
13440 0        0         1      0          0       1.07295753 -0.7218613
1412  1        0         0      0          0      -1.25293526  1.0759315
7539  0        1         0      0          0       0.67865382 -0.8061329
4621  0        1         0      0          0       0.62874196 -0.7265431
      housing_median_age population  households  median_income mean_bedrooms
2418  -0.05081113        -0.08342596 -0.50882695 -1.2394168    -0.03648780  
9990  -1.08374113         1.39212008  2.14071836 -0.7358959    -0.19291092  
13440 -0.05081113         0.28656434  0.06136148  0.1404495    -0.18700644  
1412   0.50538194         0.35897294  0.42492199  0.6344959    -0.11581168  
7539  -0.20972344         1.03802435  0.21829408 -1.0991931    -0.03247975  
4621   1.61776810         0.10024464  0.24706505 -0.6573622    -0.07763347  
      mean_rooms median_house_value
2418  -0.4145713  56700            
9990  -0.1004065 143400            
13440  0.2732884 128300            
1412   0.2741324 233200            
7539  -0.5151724 110200            
4621  -0.4598522 350900

## 1b. Cleaning - The tidyverse way! 

The code below does the same thing as the code above, but employs the tidyverse. I've pulled out the bare bones parts of Karl's 'Housing_R_tidy.r' script needed to get the data to where we want it (i.e. removed all the graphs head commands etc. Go to his original script to see the notes and visual blandishments).

I like this code more then my original version because:
1. It is easy to follow the workflow. magrittr makes it easy to see when one cleaning task ends and the next begins.
2. It is more concise.
3. The use of comments(#) after the pipes(%>%) looks professional and also makes the code more readable. Being able to share your code with others and have them understand it is very important!
4. It runs faster.

Note: in the tibble docs the function is listed as: as_tibble() not as.tibble() as first written. Oddly as.tibble() worked in my normal R deployment, but threw an error in the jupyter notebook :\ This confused me and I don't know what to make of it.

In [54]:
library(tidyverse)

housing.tidy = read_csv('housing.csv')

housing.tidy = housing.tidy %>% 
  mutate(total_bedrooms = ifelse(is.na(total_bedrooms), 
                                 median(total_bedrooms, na.rm = T),
                                 total_bedrooms),
         mean_bedrooms = total_bedrooms/households,
         mean_rooms = total_rooms/households) %>%
  select(-c(total_rooms, total_bedrooms))


categories = unique(housing.tidy$ocean_proximity) # all categories

cat_housing.tidy = categories %>% # compare the full vector against each category consecutively
  lapply(function(x) as.numeric(housing.tidy$ocean_proximity == x)) %>% # convert to numeric
  do.call("cbind", .) %>% as_tibble() # clean up
colnames(cat_housing.tidy) = categories # make nice column names

cleaned_housing.tidy = housing.tidy %>% 
  select(-c(ocean_proximity, median_house_value)) %>%
  scale() %>% as_tibble() %>%
  bind_cols(cat_housing.tidy) %>%
  add_column(median_house_value = housing.tidy$median_house_value)

set.seed(19) # Set a random seed so that same sample can be reproduced in future runs

sample = sample.int(n = nrow(cleaned_housing.tidy), size = floor(.8*nrow(cleaned_housing.tidy)), replace = F)
train = cleaned_housing.tidy[sample, ] #just the samples
test  = cleaned_housing.tidy[-sample, ] #everything but the samples
      
head(train)


Parsed with column specification:
cols(
  longitude = col_double(),
  latitude = col_double(),
  housing_median_age = col_double(),
  total_rooms = col_double(),
  total_bedrooms = col_double(),
  population = col_double(),
  households = col_double(),
  median_income = col_double(),
  median_house_value = col_double(),
  ocean_proximity = col_character()
)


longitude   latitude   housing_median_age population  households 
1  0.06473791  0.4485767 -0.05081113        -0.08342596 -0.50882695
2 -0.74882545  1.6471053 -1.08374113         1.39212008  2.14071836
3  1.07295753 -0.7218613 -0.05081113         0.28656434  0.06136148
4 -1.25293526  1.0759315  0.50538194         0.35897294  0.42492199
5  0.67865382 -0.8061329 -0.20972344         1.03802435  0.21829408
6  0.62874196 -0.7265431  1.61776810         0.10024464  0.24706505
  median_income mean_bedrooms mean_rooms NEAR BAY <1H OCEAN INLAND NEAR OCEAN
1 -1.2394168    -0.03648780   -0.4145713 0        0         1      0         
2 -0.7358959    -0.19291092   -0.1004065 0        0         1      0         
3  0.1404495    -0.18700644    0.2732884 0        0         1      0         
4  0.6344959    -0.11581168    0.2741324 1        0         0      0         
5 -1.0991931    -0.03247975   -0.5151724 0        1         0      0         
6 -0.6573622    -0.07763347   -0.4598522 0        1         0      0         
  ISLAND median_house_value
1 0       56700            
2 0      143400            
3 0      128300            
4 0      233200            
5 0      110200            
6 0      350900

## 2a. Last week's random forest model

This is the model we made at the end of last week's session I'm running it again here so that we have a benchmark to for the other models we train today.

In [4]:

########
# Random Forest Model
########
library(randomForest)
rf_model = randomForest(train_x, y = train_y , ntree = 500, importance = TRUE)

names(rf_model) #these are all the different things you can call from the model.

importance_dat = rf_model$importance
importance_dat

sorted_predictors = sort(importance_dat[,1], decreasing=TRUE)
sorted_predictors

oob_prediction = predict(rf_model) #leaving out a data source forces OOB predictions

#you may have noticed that this is avaliable using the $mse in the model options.
#but this way we learn stuff!
train_mse = mean(as.numeric((oob_prediction - train_y)^2))
oob_rmse = sqrt(train_mse)
oob_rmse


y_pred_rf = predict(rf_model , test_x)
test_mse = mean(((y_pred_rf - test_y)^2))
test_rmse = sqrt(test_mse)
test_rmse # ~48620



randomForest 4.6-12
Type rfNews() to see new features/changes/bug fixes.

Attaching package: ‘randomForest’

The following object is masked from ‘package:dplyr’:

    combine

The following object is masked from ‘package:ggplot2’:

    margin



[1] "call"            "type"            "predicted"       "mse"            
 [5] "rsq"             "oob.times"       "importance"      "importanceSD"   
 [9] "localImportance" "proximity"       "ntree"           "mtry"           
[13] "forest"          "coefs"           "y"               "test"           
[17] "inbag"

%IncMSE      IncNodePurity
NEAR BAY           4.450369e+08 1.391014e+12 
<1H OCEAN          1.427100e+09 4.338351e+12 
INLAND             3.703060e+09 3.116839e+13 
NEAR OCEAN         4.262002e+08 2.119833e+12 
ISLAND             6.761432e+04 1.682004e+10 
longitude          6.686360e+09 2.580455e+13 
latitude           5.375308e+09 2.235183e+13 
housing_median_age 1.062666e+09 9.808021e+12 
population         1.081431e+09 7.441795e+12 
households         1.193537e+09 7.897400e+12 
median_income      8.325735e+09 7.167956e+13 
mean_bedrooms      4.083603e+08 7.597531e+12 
mean_rooms         1.890830e+09 2.155527e+13

median_income          longitude           latitude             INLAND 
      8.325735e+09       6.686360e+09       5.375308e+09       3.703060e+09 
        mean_rooms          <1H OCEAN         households         population 
      1.890830e+09       1.427100e+09       1.193537e+09       1.081431e+09 
housing_median_age           NEAR BAY         NEAR OCEAN      mean_bedrooms 
      1.062666e+09       4.450369e+08       4.262002e+08       4.083603e+08 
            ISLAND 
      6.761432e+04

[1] 48973.79

[1] 48392.87

$48392 is the test error benchmark based off that random forest run.

## 2b. Gradient Boosting

Gradient boosting is an ensemble supervised machine learning model that builds up the concept of the random forest algorithm we explored last week. Recall that for a random forest we spawned 500 decision trees and took the mean of their predictions to get a 'wisdom of the crowd' effect and arrive at a more accurate prediction than any one tree would provide.

Here we use the Extreme Gradient Boosting library to implement this in R. Note the 'Extreme' in extreme gradient boosting refers to the computational efficiency. The algorithm could more accurately could be described as 'regularized gradient boosting'.

###  Gradient Boosting - Algorithm details

Extreme gradient boosting also builds a forest of trees, but does so in an additive manner. The algorithm iteratively builds trees that minimize the error, and thereby descends towards an optimal set of predictive trees. Already learned trees are kept, and new trees are added one after another to minimize the objective function (error in predictions). The trees are grown sequentially: each tree is grown using information from previously grown trees. Each tree is fit on a modified version of the original data set based on the previous trees built.

The trees are accompanied by a regularization paramater to avoid overfit.  

One difference between boosting and random forests: in boosting, because the growth of a particular 
tree takes into account the other trees that have already been grown, smaller trees are typically sufficient (less splits and depth). 

In [10]:
######
# XG Boost
######
# see the docs: http://cran.fhcrc.org/web/packages/xgboost/vignettes/xgboost.pdf
library(xgboost)

#put into the xgb matrix format
dtrain = xgb.DMatrix(data =  as.matrix(train_x), label = train_y )
dtest = xgb.DMatrix(data =  as.matrix(test_x), label = test_y)

# these are the datasets the rmse is evaluated for at each iteration
watchlist = list(train=dtrain, test=dtest)

# try 1 - off a set of paramaters I know work pretty well for most stuff

bst = xgb.train(data = dtrain, 
                max.depth = 8, 
                eta = 0.3, 
                nthread = 2, 
                nround = 1000, 
                watchlist = watchlist, 
                objective = "reg:linear", 
                early_stopping_rounds = 50)

[1]	train-rmse:171405.234375	test-rmse:173186.609375 
Multiple eval metrics are present. Will use test_rmse for early stopping.
Will train until test_rmse hasn't improved in 50 rounds.

[2]	train-rmse:126655.414062	test-rmse:129169.820312 
[3]	train-rmse:96442.976562	test-rmse:100198.515625 
[4]	train-rmse:76786.750000	test-rmse:81659.406250 
[5]	train-rmse:64022.230469	test-rmse:70192.492188 
[6]	train-rmse:55857.519531	test-rmse:63414.535156 
[7]	train-rmse:50131.742188	test-rmse:58767.277344 
[8]	train-rmse:46460.761719	test-rmse:56293.750000 
[9]	train-rmse:43681.136719	test-rmse:54576.554688 
[10]	train-rmse:41670.960938	test-rmse:53205.472656 
[11]	train-rmse:40081.183594	test-rmse:52478.437500 
[12]	train-rmse:38772.078125	test-rmse:51718.906250 
[13]	train-rmse:37653.445312	test-rmse:51174.562500 
[14]	train-rmse:36709.722656	test-rmse:50757.242188 
[15]	train-rmse:35945.863281	test-rmse:50291.871094 
[16]	train-rmse:35438.671875	test-rmse:50195.484375 
[17]	train-rmse:34626.86

So our first run there gets a rmse of $47723, an improvement over our benchmark model. That isn't the end of the story though, we can try to squeak out further improvements through 'hyperparameter tuning'. A hyperparameter is one of the mutable options that we pass to the algorithm along with our data.

## 2c. Tuning the algorithm - hyperparameters for xgboost

Boosting has 3 tuning paramaters that we can focus on

1. The number of trees. Here we use a good trick, instead of specifying an exact number, we give the algorithm a big number (nround = 10000) and the param (early_stopping_rounds = 50). This effectively means: 'keep iteratively growing trees until you have 10,000 of them, or stop early if the scores haven't improved for the last 50 rounds'.
2. The shrinkage parameter λ (eta in the params), a small positive number. This controls the rate at which boosting learns. Typical values are 0.01 or 0.001, and the right choice can depend on the problem. Very small λ can require using a very large value of B in order to achieve good performance.
3. The number of splits in each tree, which controls the complexity of the boosted ensemble (controlled with max.depth).

### Here we try a 'slower learning' model. The up and down weights for each iteration are smaller we also use more iterations to account for the fact that the model will take longer to learn.

In [13]:
bst_slow = xgb.train(data = dtrain, 
                        max.depth=5, 
                        eta = 0.01, 
                        nthread = 2, 
                        nround = 10000, 
                        watchlist = watchlist, 
                        objective = "reg:linear", 
                        early_stopping_rounds = 50)


[1]	train-rmse:234537.453125	test-rmse:235377.093750 
Multiple eval metrics are present. Will use test_rmse for early stopping.
Will train until test_rmse hasn't improved in 50 rounds.

[2]	train-rmse:232401.531250	test-rmse:233245.140625 
[3]	train-rmse:230289.125000	test-rmse:231137.718750 
[4]	train-rmse:228199.281250	test-rmse:229051.968750 
[5]	train-rmse:226132.109375	test-rmse:226991.000000 
[6]	train-rmse:224087.062500	test-rmse:224951.265625 
[7]	train-rmse:222062.828125	test-rmse:222929.796875 
[8]	train-rmse:220062.687500	test-rmse:220935.281250 
[9]	train-rmse:218083.812500	test-rmse:218961.875000 
[10]	train-rmse:216124.953125	test-rmse:217008.671875 
[11]	train-rmse:214189.453125	test-rmse:215078.078125 
[12]	train-rmse:212273.171875	test-rmse:213167.406250 
[13]	train-rmse:210378.359375	test-rmse:211287.390625 
[14]	train-rmse:208504.187500	test-rmse:209421.609375 
[15]	train-rmse:206646.906250	test-rmse:207576.968750 
[16]	train-rmse:204814.375000	test-rmse:205758.32812

In [14]:
rf_benchmark = 48392

bst_slow$best_score / rf_benchmark

test-rmse 
0.9345753

Things to note:
- 6009 iterations were run, and it backtracked to 5959 to get the best one thanks to our early stopping rounds parameter.
- So that is an improvement of ~6.5% in rmse error over last week. Yay lets go home. Wait! What we have done here is fit to the training set and the training set at the same time (leading to model overfit).  

## Problems to address

### 1. We need to work with a validation set and only at the end evaluate the model performance against the test set.
Remember our test set should be withheld to evaluate the model on data it hasn't seen. By iteratively checking the train and test rmse above, we have violated that rule and so we cannot say how accurate our model is on data it hasn't seen. To do this we need a validation set (essentially a second test set) that the model can peek at after each iteration to see how well it is performing or external data, then with the final version we can make the assessment vs. the test set.

### 2 .If we kept tweaking one hyperparameter, waiting to see the result and then repeating the process we will be here forever. We need to speed this up in a systematic fashion


## 2d. A validation set

validation set - Another subset of our data that is witheld from the training algorithm, but compared against at each iteration to see how the model is performing.

Here we make this through the same method as the test set, by sampling 20% of the remaining training set and passing this into the xgboost watchlist. The algorithm will watch the rmse of the training (which it can learn from) and the validation (which it can't learn parameters from, only see the rmse outcome) and continue until one is no longer improving


### Note on tidyverse
train_y = train_t[,'median_house_value']
When I run the tidy cleaning for step one this outputs a dataframe like column, but the algorithm needs a plain vector of numbers for the y labels.

The code needs to switch to:

train_y = pull(train_t, median_house_value)

In order to un-tibble the data

In [76]:
####
# Proper use - validation set
####


sample = sample.int(n = nrow(train), size = floor(.8*nrow(train)), replace = F)

train_t = train[sample, ] #just the samples
valid  = train[-sample, ] #everything but the samples

train_y = train_t[,'median_house_value']

#if tidyverse was used, dplyr pull function solves the problem:
#train_y = pull(train_t, median_house_value)
train_x = train_t[, names(train_t) !='median_house_value']

valid_y = valid[,'median_house_value']
valid_x = valid[, names(train_t) !='median_house_value']

train_y

[1] 259400 136800 151800 249400  71900  92300  57000 354200  89200  57400
   [11]  85400 202900 165200 118800 164500 229800  96900 337600 114700 118800
   [21] 169200 162900 160700 274800 500001  93400 269400 253300 114200 237500
   [31] 126200 225000 355900 169400  62800 137500 195900 112500 229900 164300
   [41] 176100 500001 255400 307400 243800 157300  98300 338400 225000 223900
   [51] 380000 285500 487800  97600 173100 331200 262500 133300 174500 146900
   [61]  62700  78000 326100 173600  52500 258100  78800 162500 234200 423700
   [71] 287700 387300 260100  92500 288900  94400 112500  75000 283300  46700
   [81] 137000 230600 382500  60800 103000  93700 160200  69700 248700 100000
   [91] 348300 417100  77100 148800 181000 218800  99100 107200 364700 194100
  [101] 218200 365900 495500 307400  85600 192800  71500 171300  72700 194800
  [111] 116900 322600  71300 500001 500001 225600 294300 153500 138100 155500
  [121]  80600 500001 139200 185900 376100  53500 241400 184900 118400 122700
  [131] 430100 410700 196100 152100 397500 284700 111600 108800 398200 467600
  [141] 187700 455400  53100 183800 106300 188000  96100 197400  75000 319700
  [151] 500001 101600 230700 241700 213300 372000  98400 257800 145800  90500
  [161] 188400 236900 174800 162500 396400 191000 240900 371300 306700 247200
  [171] 230400 164700 347800 100000 286500  91400 319100 465800 162500 169000
  [181] 271800  92400 113900 111600 232100  96100 445400 159500 117300 170800
  [191]  85700 200000 181800 278100 405200 255800 187500 114500  83300 190000
  [201] 193500 117600  92400 224400 230000 350000 153300 122900 132600 148700
  [211] 136600  91500 166800 254100 109100 368000  82800 190700 203600 264400
  [221] 273600 225800  81700 500000 171000 160900 336600 500001 500001 144000
  [231]  58700 351100 112500  77900  43000  83200  85400 301700  63000  89500
  [241] 282100 191900  98600  97900 264900  70300 367100 500001 177500  65600
  [251] 153000 226800 101600 196000 263000 290500  81100 157700 255000 172100
  [261] 260700 275400  91200 341300  98600 181300  56800  57700 238300 140600
  [271]  91400 184100 169300 250000 186900 218100  76100 224900 477100 150000
  [281] 163800 360100 272800 500001 405000 332600  63500 287900 187800 100200
  [291] 100600  69600 243300  45000 107100  74800 178800 389500 154500  77000
  [301] 212500  68100 226300  93800 168900  65400 287500 142600  86300  83800
  [311] 242000  52700 442100 500001  97800  93100 162500  94600 111800 162700
  [321] 133800 189700 418400 128300 145800 106300 280300 126500 155600 154800
  [331] 140000 158800 188900  72600 225000 167400  81500 134100 209800 232600
  [341] 245200 224100 297500 148300 500001  59300 270100  64700  99500 243000
  [351] 196200  63500 185200  55000  70700 381800 284100 237900 500001 204300
  [361] 199800 149000 203400 233800 150200  62700  64400  92600 478400 160100
  [371] 269900 241600  86400 175900 335200 194800  47700 209700 120800 134500
  [381] 350000 101300 139000 160700 426400 113400 324700  65100  66800  90000
  [391] 361900 500001 175300 157300 182700 500001 388100  75500 271800 133000
  [401] 158400  69100 500001  65400 307100 103100 221100 365500 381300 226500
  [411] 296400 320300 224200 264200  77700  70900 500001 388300 132000  41300
  [421]  97300 446000 269800  73400 231300  98400 209100  88900  57000 500001
  [431] 160700 218200 224700 146800 289700 500001 218300  93000 302100 363300
  [441] 262300  62000 141600  43700 143400 139600 274600 124700 175000 178900
  [451] 101800 435900 500001 192200 233300 472700 252700 152900 167500 339600
  [461] 356100 200000 209600 331800  68400 162200 179300 500001 176700  59200
  [471]  85800 233100 134900  22500 238500 132000 433300 150000 206400 329400
  [481] 169900 135400 129500 140400 210000 162500 101600 196000  69800 156700
  [491] 243300 205600 114600  79400  83200 500001 158900 254400 247700 153900
  [501] 195700 154200  71500 387100 202800 112500 470800 165600 287600 147800
  [511] 150300 146200  619

In [77]:
gb_train = xgb.DMatrix(data = as.matrix(train_x), label = train_y )
gb_valid = xgb.DMatrix(data = as.matrix(valid_x), label = valid_y )
#in jupyter the label needs to be in an as.matrix() or I get an error? subtle and annoying differences

# train xgb, evaluating against the validation
watchlist = list(train = gb_train, valid = gb_valid)


We then run the xgboost algorithm again and after training we evaluate on the test data.

In [51]:
bst_slow = xgb.train(data= gb_train, 
                        max.depth = 10, 
                        eta = 0.01, 
                        nthread = 2, 
                        nround = 10000, 
                        watchlist = watchlist, 
                        objective = "reg:linear", 
                        early_stopping_rounds = 50)

[1]	train-rmse:234516.140625	valid-rmse:234365.062500 
Multiple eval metrics are present. Will use valid_rmse for early stopping.
Will train until valid_rmse hasn't improved in 50 rounds.

[2]	train-rmse:232324.312500	valid-rmse:232196.203125 
[3]	train-rmse:230154.421875	valid-rmse:230052.109375 
[4]	train-rmse:228008.031250	valid-rmse:227934.031250 
[5]	train-rmse:225883.390625	valid-rmse:225838.093750 
[6]	train-rmse:223780.015625	valid-rmse:223764.781250 
[7]	train-rmse:221698.312500	valid-rmse:221711.281250 
[8]	train-rmse:219638.015625	valid-rmse:219679.843750 
[9]	train-rmse:217598.515625	valid-rmse:217662.437500 
[10]	train-rmse:215579.500000	valid-rmse:215679.562500 
[11]	train-rmse:213581.296875	valid-rmse:213699.109375 
[12]	train-rmse:211604.843750	valid-rmse:211760.078125 
[13]	train-rmse:209648.484375	valid-rmse:209831.375000 
[14]	train-rmse:207712.937500	valid-rmse:207932.953125 
[15]	train-rmse:205795.906250	valid-rmse:206053.218750 
[16]	train-rmse:203899.781250	valid

In [52]:

# recall we ran the following to get the test data in the right format:
# dtest = xgb.DMatrix(data =  as.matrix(test_x), label = test_y)
# here I have it with the label taken off, just to remind us its external data xgb would ignore the label though during predictions
dtest = xgb.DMatrix(data =  as.matrix(test_x))

#test the model on truly external data

y_hat_valid = predict(bst_slow, dtest)

test_mse = mean(((y_hat_valid - test_y)^2))
test_rmse = sqrt(test_mse)
test_rmse 


[1] 46793.21

In [53]:
test_rmse/rf_benchmark

[1] 0.9669617

test-rmse: 46793.21
This is higher then on the first run, but we can be confident that the improved score is not due to overfit thanks to our use of a validation set! 

A lower rmse isn't necessarily better, as we now have more confidence in external predictions. 
This ansewer give an ~3.5% improvement over a basic random forest... is it worth the effort? The answer to this question always depends on the purpose of the model.


## 3a. Grid Search to find the best hyperparameter combinations

In [ ]:

###
# Grid search first principles 
###

max.depths = c(3, 5, 7, 9)
etas = c(0.01, 0.001, 0.0001)

best_params = 0
best_score = 0

count = 1
for( depth in max.depths ){
    for( num in etas){

        bst_grid = xgb.train(data = gb_train, 
                                max.depth = depth, 
                                eta=num, 
                                nthread = 2, 
                                nround = 10000, 
                                watchlist = watchlist, 
                                objective = "reg:linear", 
                                early_stopping_rounds = 50, 
                                verbose=0)

        if(count == 1){
            best_params = bst_grid$params
            best_score = bst_grid$best_score
            count = count + 1
            }
        else if( bst_grid$best_score < best_score){
            best_params = bst_grid$params
            best_score = bst_grid$best_score
        }
    }
}

best_params
best_score
#valid-rmse: 47033.28

# max_depth of 9, eta of 0.01
bst_tuned = xgb.train( data = gb_train, 
                        max.depth = 9, 
                        eta = 0.01, 
                        nthread = 2, 
                        nround = 10000, 
                        watchlist = watchlist, 
                        objective = "reg:linear", 
                        early_stopping_rounds = 50)

y_hat_xgb_grid = predict(bst_tuned, dtest)

test_mse = mean(((y_hat_xgb_grid - test_y)^2))
test_rmse = sqrt(test_mse)
test_rmse # test-rmse: 46675
# By tuning the hyperparamaters we have moved to a 4% improvement over random forest




In [ ]:

#######
# tweak the hyperparamaters using a grid search
# The caret package (short for classification and regression training)
######

library(caret) 

# look up the model we are running to see the paramaters
modelLookup("xgbLinear")
 
# set up all the pairwise combinations

xgb_grid_1 = expand.grid(nrounds = c(1000,2000,3000,4000) ,
                            eta = c(0.01, 0.001, 0.0001),
                            lambda = 1,
                            alpha = 0)
xgb_grid_1


#here we do one better then a validation set, we use cross validation to 
#expand the amount of info we have!
xgb_trcontrol_1 = trainControl(method = "cv",
                                number = 5,
                                verboseIter = TRUE,
                                returnData = FALSE,
                                returnResamp = "all", 
                                allowParallel = TRUE)


######
#below a grid-search, cross-validation xgboost model in caret
######
# train the model for each parameter combination in the grid, using CV to evaluate on multiple folds. Make sure your laptop is plugged in or RIP battery.

# note how this is now the caret train function
?train

xgb_train_1 = train(x = as.matrix(train_x),
                    y = train_y,
                    trControl = xgb_trcontrol_1,
                    tuneGrid = xgb_grid_1,
                    method = "xgbLinear",
                    max.depth = 5)

names(xgb_train_1)
xgb_train_1$bestTune
xgb_train_1$method
summary(xgb_train_1)


#alternatively, you can 'narrow in' on the best paramaters by taking a range of options around the best values found and seeing if high resolution tweaks can provide even further improvements.

xgb_cv_yhat = predict(xgb_train_1 , as.matrix(test_x))


test_mse = mean(((xgb_cv_yhat - test_y)^2))
test_rmse = sqrt(test_mse)
test_rmse # 46641... pretty close to the 'by hand' grid search!

#Cam's hypothesis - not using 'early stopping rounds' here so the model isn't cutting out at the exact best point. re-running this with a validation setup as opposed to a cv setup would allow us to implement a grid search efficiently and wind up with the best hyperparamaters. I shall leave this as a follow up exercise for the curious.



In [ ]:

########
# Ensemble the models together, 
# strategy for when accuracy is more important then knowing the best predictors
########


y_pred_rf #random forest
y_hat_valid #xgBoost with validation
y_hat_xgb_grid #xgBoost grid search
xgb_cv_yhat #xgBoost caret cross validation

length(y_hat_xgb_grid)


blend_pred = (y_hat * .25) + (y_pred_rf * .25) + (xgb_cv_yhat * .25) + (y_hat_xgb_grid * .25)
length(blend_pred)

length(blend_pred) == length(y_hat_xgb_grid)

blend_test_mse = mean(((blend_pred - test_y)^2))
blend_test_rmse = sqrt(blend_test_mse)
blend_test_rmse # 45205 by averaging just 4 predictors we have dropped the rmse a few percent lower then the best scoring of the 4 models. This does come at a cost though, we now can't make accurate inferrences about the best predictors!

#next step - you can grid search the weights of the ensemble to try and drop the rmse further!

